File: src/glottisdale/speak/phonetic_distance.py
"""ARPABET phonetic feature matrix and distance calculations."""

# IPA-to-ARPABET mapping for phonemes produced by BFA aligner.
# Diphthongs first (multi-char), then monophthongs, then consonants.
IPA_TO_ARPABET: dict[str, str] = {
    # Diphthongs (must be checked before single-char vowels)
    "aɪ": "AY",
    "aʊ": "AW",
    "eɪ": "EY",
    "oʊ": "OW",
    "ɔɪ": "OY",
    # Vowels
    "i":  "IY",
    "ɪ":  "IH",
    "e":  "EY",
    "ɛ":  "EH",
    "æ":  "AE",
    "ɑ":  "AA",
    "ɒ":  "AA",
    "ɔ":  "AO",
    "o":  "OW",
    "ʊ":  "UH",
    "u":  "UW",
    "ə":  "AH",
    "ɜ":  "ER",
    "ɐ":  "AH",
    "ʌ":  "AH",
    "a":  "AA",
    # Consonants — stops
    "p":  "P",
    "b":  "B",
    "t":  "T",
    "d":  "D",
    "k":  "K",
    "g":  "G",
    # Consonants — nasals
    "m":  "M",
    "n":  "N",
    "ŋ":  "NG",
    "ɲ":  "N",
    "ɴ":  "NG",
    # Consonants — fricatives
    "f":  "F",
    "v":  "V",
    "θ":  "TH",
    "ð":  "DH",
    "s":  "S",
    "z":  "Z",
    "ʃ":  "SH",
    "ʒ":  "ZH",
    "h":  "HH",
    "ɦ":  "HH",
    "ç":  "HH",
    "x":  "HH",
    "ɣ":  "G",
    # Consonants — liquids/rhotics
    "l":  "L",
    "ɫ":  "L",
    "ɬ":  "L",
    "ɮ":  "L",
    "r":  "R",
    "ɹ":  "R",
    "ɾ":  "R",
    "ɽ":  "R",
    "ʁ":  "R",
    "ʀ":  "R",
    # Consonants — glides
    "j":  "Y",
    "w":  "W",
    "ɥ":  "W",
}

# Ordered list of multi-char IPA keys for prefix matching
_IPA_DIPHTHONGS = sorted(
    [k for k in IPA_TO_ARPABET if len(k) > 1],
    key=len, reverse=True,
)


def normalize_phoneme(phoneme: str) -> str:
    """Convert an IPA phoneme to ARPABET if possible, passthrough otherwise.

    Strips length markers (ː, ˑ) before lookup. Checks multi-character
    diphthongs first, then single characters.
    """
    if not phoneme:
        return phoneme

    # Already ARPABET (uppercase letters, possibly with stress digit)
    base = phoneme.rstrip("012")
    if base and base.isascii() and base.isupper():
        return phoneme

    # Strip IPA length markers
    cleaned = phoneme.rstrip("ːˑ")

    # Try multi-char diphthongs first
    for diph in _IPA_DIPHTHONGS:
        if cleaned.startswith(diph):
            return IPA_TO_ARPABET[diph]

    # Try single-char lookup
    if cleaned in IPA_TO_ARPABET:
        return IPA_TO_ARPABET[cleaned]

    # Unknown — return as-is
    return phoneme


# Articulatory features for each ARPABET phoneme (stress-stripped).
# Consonants: (type, manner, place, voicing)
# Vowels: (type, height, backness, roundness, tenseness)
# "type" distinguishes consonants from vowels for max-distance fallback.

FEATURES: dict[str, tuple[str, ...]] = {
    # Consonants — (type, manner, place, voicing)
    "P":  ("consonant", "stop",      "bilabial",      "voiceless"),
    "B":  ("consonant", "stop",      "bilabial",      "voiced"),
    "T":  ("consonant", "stop",      "alveolar",      "voiceless"),
    "D":  ("consonant", "stop",      "alveolar",      "voiced"),
    "K":  ("consonant", "stop",      "velar",         "voiceless"),
    "G":  ("consonant", "stop",      "velar",         "voiced"),
    "F":  ("consonant", "fricative", "labiodental",   "voiceless"),
    "V":  ("consonant", "fricative", "labiodental",   "voiced"),
    "TH": ("consonant", "fricative", "dental",        "voiceless"),
    "DH": ("consonant", "fricative", "dental",        "voiced"),
    "S":  ("consonant", "fricative", "alveolar",      "voiceless"),
    "Z":  ("consonant", "fricative", "alveolar",      "voiced"),
    "SH": ("consonant", "fricative", "postalveolar",  "voiceless"),
    "ZH": ("consonant", "fricative", "postalveolar",  "voiced"),
    "HH": ("consonant", "fricative", "glottal",       "voiceless"),
    "CH": ("consonant", "affricate", "postalveolar",  "voiceless"),
    "JH": ("consonant", "affricate", "postalveolar",  "voiced"),
    "M":  ("consonant", "nasal",     "bilabial",      "voiced"),
    "N":  ("consonant", "nasal",     "alveolar",      "voiced"),
    "NG": ("consonant", "nasal",     "velar",         "voiced"),
    "L":  ("consonant", "liquid",    "alveolar",      "voiced"),
    "R":  ("consonant", "liquid",    "postalveolar",  "voiced"),
    "W":  ("consonant", "glide",     "bilabial",      "voiced"),
    "Y":  ("consonant", "glide",     "palatal",       "voiced"),
    # Vowels — (type, height, backness, roundness, tenseness)
    "IY": ("vowel", "high",  "front",   "unrounded", "tense"),
    "IH": ("vowel", "high",  "front",   "unrounded", "lax"),
    "EY": ("vowel", "mid",   "front",   "unrounded", "tense"),
    "EH": ("vowel", "mid",   "front",   "unrounded", "lax"),
    "AE": ("vowel", "low",   "front",   "unrounded", "lax"),
    "AA": ("vowel", "low",   "back",    "unrounded", "tense"),
    "AH": ("vowel", "mid",   "central", "unrounded", "lax"),
    "AO": ("vowel", "mid",   "back",    "rounded",   "tense"),
    "OW": ("vowel", "mid",   "back",    "rounded",   "tense"),
    "UH": ("vowel", "high",  "back",    "rounded",   "lax"),
    "UW": ("vowel", "high",  "back",    "rounded",   "tense"),
    "AW": ("vowel", "low",   "central", "unrounded", "tense"),
    "AY": ("vowel", "low",   "central", "unrounded", "tense"),
    "OY": ("vowel", "mid",   "back",    "rounded",   "tense"),
    "ER": ("vowel", "mid",   "central", "rounded",   "tense"),
}

# Max distance when comparing a consonant to a vowel (different type)
_CROSS_TYPE_DISTANCE = 5


def strip_stress(phoneme: str) -> str:
    """Remove trailing stress marker (0, 1, 2) from an ARPABET phoneme."""
    if phoneme and phoneme[-1] in "012":
        return phoneme[:-1]
    return phoneme


def phoneme_distance(a: str, b: str) -> int:
    """Compute articulatory feature distance between two ARPABET phonemes.

    Stress markers are ignored. Returns 0 for identical phonemes,
    higher values for more dissimilar phonemes.
    """
    a_base = strip_stress(a)
    b_base = strip_stress(b)

    if a_base == b_base:
        return 0

    feat_a = FEATURES.get(a_base)
    feat_b = FEATURES.get(b_base)

    if feat_a is None or feat_b is None:
        return _CROSS_TYPE_DISTANCE

    # Different type (consonant vs vowel) = max distance
    if feat_a[0] != feat_b[0]:
        return _CROSS_TYPE_DISTANCE

    # Count differing features (skip index 0 which is the type tag)
    return sum(1 for fa, fb in zip(feat_a[1:], feat_b[1:]) if fa != fb)


def syllable_distance(a_phonemes: list[str], b_phonemes: list[str]) -> int:
    """Compute distance between two syllables (lists of ARPABET phonemes).

    Aligns by padding the shorter list with None, which incurs max distance
    per missing phoneme.
    """
    len_a = len(a_phonemes)
    len_b = len(b_phonemes)
    max_len = max(len_a, len_b)

    if max_len == 0:
        return 0

    total = 0
    for i in range(max_len):
        pa = a_phonemes[i] if i < len_a else None
        pb = b_phonemes[i] if i < len_b else None
        if pa is None or pb is None:
            total += _CROSS_TYPE_DISTANCE
        else:
            total += phoneme_distance(pa, pb)
    return total

File: crates/core/src/speak/phonetic_distance.rs
//! ARPABET phonetic feature matrix and distance calculations.

use std::collections::HashMap;

lazy_static::lazy_static! {
    /// IPA-to-ARPABET mapping for phonemes produced by BFA aligner.
    static ref IPA_TO_ARPABET: Vec<(&'static str, &'static str)> = vec![
        // Diphthongs first (multi-char)
        ("aɪ", "AY"), ("aʊ", "AW"), ("eɪ", "EY"), ("oʊ", "OW"), ("ɔɪ", "OY"),
        // Vowels
        ("i", "IY"), ("ɪ", "IH"), ("e", "EY"), ("ɛ", "EH"), ("æ", "AE"),
        ("ɑ", "AA"), ("ɒ", "AA"), ("ɔ", "AO"), ("o", "OW"), ("ʊ", "UH"),
        ("u", "UW"), ("ə", "AH"), ("ɜ", "ER"), ("ɐ", "AH"), ("ʌ", "AH"),
        ("a", "AA"),
        // Consonants — stops
        ("p", "P"), ("b", "B"), ("t", "T"), ("d", "D"), ("k", "K"), ("g", "G"),
        // Consonants — nasals
        ("m", "M"), ("n", "N"), ("ŋ", "NG"), ("ɲ", "N"), ("ɴ", "NG"),
        // Consonants — fricatives
        ("f", "F"), ("v", "V"), ("θ", "TH"), ("ð", "DH"), ("s", "S"),
        ("z", "Z"), ("ʃ", "SH"), ("ʒ", "ZH"), ("h", "HH"), ("ɦ", "HH"),
        ("ç", "HH"), ("x", "HH"), ("ɣ", "G"),
        // Consonants — liquids/rhotics
        ("l", "L"), ("ɫ", "L"), ("ɬ", "L"), ("ɮ", "L"),
        ("r", "R"), ("ɹ", "R"), ("ɾ", "R"), ("ɽ", "R"), ("ʁ", "R"), ("ʀ", "R"),
        // Consonants — glides
        ("j", "Y"), ("w", "W"), ("ɥ", "W"),
    ];

    /// Articulatory features for each ARPABET phoneme.
    static ref FEATURES: HashMap<&'static str, &'static [&'static str]> = {
        let mut m = HashMap::new();
        // Consonants: [type, manner, place, voicing]
        m.insert("P",  &["consonant", "stop", "bilabial", "voiceless"][..]);
        m.insert("B",  &["consonant", "stop", "bilabial", "voiced"][..]);
        m.insert("T",  &["consonant", "stop", "alveolar", "voiceless"][..]);
        m.insert("D",  &["consonant", "stop", "alveolar", "voiced"][..]);
        m.insert("K",  &["consonant", "stop", "velar", "voiceless"][..]);
        m.insert("G",  &["consonant", "stop", "velar", "voiced"][..]);
        m.insert("F",  &["consonant", "fricative", "labiodental", "voiceless"][..]);
        m.insert("V",  &["consonant", "fricative", "labiodental", "voiced"][..]);
        m.insert("TH", &["consonant", "fricative", "dental", "voiceless"][..]);
        m.insert("DH", &["consonant", "fricative", "dental", "voiced"][..]);
        m.insert("S",  &["consonant", "fricative", "alveolar", "voiceless"][..]);
        m.insert("Z",  &["consonant", "fricative", "alveolar", "voiced"][..]);
        m.insert("SH", &["consonant", "fricative", "postalveolar", "voiceless"][..]);
        m.insert("ZH", &["consonant", "fricative", "postalveolar", "voiced"][..]);
        m.insert("HH", &["consonant", "fricative", "glottal", "voiceless"][..]);
        m.insert("CH", &["consonant", "affricate", "postalveolar", "voiceless"][..]);
        m.insert("JH", &["consonant", "affricate", "postalveolar", "voiced"][..]);
        m.insert("M",  &["consonant", "nasal", "bilabial", "voiced"][..]);
        m.insert("N",  &["consonant", "nasal", "alveolar", "voiced"][..]);
        m.insert("NG", &["consonant", "nasal", "velar", "voiced"][..]);
        m.insert("L",  &["consonant", "liquid", "alveolar", "voiced"][..]);
        m.insert("R",  &["consonant", "liquid", "postalveolar", "voiced"][..]);
        m.insert("W",  &["consonant", "glide", "bilabial", "voiced"][..]);
        m.insert("Y",  &["consonant", "glide", "palatal", "voiced"][..]);
        // Vowels: [type, height, backness, roundness, tenseness]
        m.insert("IY", &["vowel", "high", "front", "unrounded", "tense"][..]);
        m.insert("IH", &["vowel", "high", "front", "unrounded", "lax"][..]);
        m.insert("EY", &["vowel", "mid", "front", "unrounded", "tense"][..]);
        m.insert("EH", &["vowel", "mid", "front", "unrounded", "lax"][..]);
        m.insert("AE", &["vowel", "low", "front", "unrounded", "lax"][..]);
        m.insert("AA", &["vowel", "low", "back", "unrounded", "tense"][..]);
        m.insert("AH", &["vowel", "mid", "central", "unrounded", "lax"][..]);
        m.insert("AO", &["vowel", "mid", "back", "rounded", "tense"][..]);
        m.insert("OW", &["vowel", "mid", "back", "rounded", "tense"][..]);
        m.insert("UH", &["vowel", "high", "back", "rounded", "lax"][..]);
        m.insert("UW", &["vowel", "high", "back", "rounded", "tense"][..]);
        m.insert("AW", &["vowel", "low", "central", "unrounded", "tense"][..]);
        m.insert("AY", &["vowel", "low", "central", "unrounded", "tense"][..]);
        m.insert("OY", &["vowel", "mid", "back", "rounded", "tense"][..]);
        m.insert("ER", &["vowel", "mid", "central", "rounded", "tense"][..]);
        m
    };
}

const CROSS_TYPE_DISTANCE: i32 = 5;

/// Strip trailing stress marker (0, 1, 2) from an ARPABET phoneme.
pub fn strip_stress(phoneme: &str) -> &str {
    phoneme.trim_end_matches(|c: char| c.is_ascii_digit())
}

/// Convert an IPA phoneme to ARPABET if possible, passthrough otherwise.
pub fn normalize_phoneme(phoneme: &str) -> String {
    if phoneme.is_empty() {
        return phoneme.to_string();
    }

    // Already ARPABET (uppercase ASCII, possibly with stress digit)
    let base = strip_stress(phoneme);
    if !base.is_empty() && base.is_ascii() && base.chars().all(|c| c.is_uppercase()) {
        return phoneme.to_string();
    }

    // Strip IPA length markers
    let cleaned = phoneme.trim_end_matches(['ː', 'ˑ']);

    // Try multi-char diphthongs first, then single-char
    for (ipa, arpabet) in IPA_TO_ARPABET.iter() {
        if cleaned.starts_with(ipa) {
            return arpabet.to_string();
        }
    }

    phoneme.to_string()
}

/// Compute articulatory feature distance between two ARPABET phonemes.
///
/// Stress markers are ignored. Returns 0 for identical phonemes.
pub fn phoneme_distance(a: &str, b: &str) -> i32 {
    let a_base = strip_stress(a);
    let b_base = strip_stress(b);

    if a_base == b_base {
        return 0;
    }

    let feat_a = FEATURES.get(a_base);
    let feat_b = FEATURES.get(b_base);

    match (feat_a, feat_b) {
        (Some(fa), Some(fb)) => {
            if fa[0] != fb[0] {
                return CROSS_TYPE_DISTANCE;
            }
            fa[1..].iter().zip(fb[1..].iter()).filter(|(a, b)| a != b).count() as i32
        }
        _ => CROSS_TYPE_DISTANCE,
    }
}

/// Compute distance between two syllables (lists of ARPABET phonemes).
pub fn syllable_distance(a: &[String], b: &[String]) -> i32 {
    let max_len = a.len().max(b.len());
    if max_len == 0 {
        return 0;
    }

    let mut total = 0;
    for i in 0..max_len {
        match (a.get(i), b.get(i)) {
            (Some(pa), Some(pb)) => total += phoneme_distance(pa, pb),
            _ => total += CROSS_TYPE_DISTANCE,
        }
    }
    total
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_phoneme_distance_identical() {
        assert_eq!(phoneme_distance("K", "K"), 0);
        assert_eq!(phoneme_distance("AE1", "AE0"), 0); // stress ignored
    }

    #[test]
    fn test_phoneme_distance_same_type() {
        // P and B: same manner, same place, different voicing = 1
        assert_eq!(phoneme_distance("P", "B"), 1);
        // P and K: same manner, same voicing, different place = 1
        assert_eq!(phoneme_distance("P", "K"), 1);
    }

    #[test]
    fn test_phoneme_distance_cross_type() {
        // Consonant vs vowel = max distance
        assert_eq!(phoneme_distance("K", "AE1"), CROSS_TYPE_DISTANCE);
    }

    #[test]
    fn test_phoneme_distance_unknown() {
        assert_eq!(phoneme_distance("K", "UNKNOWN"), CROSS_TYPE_DISTANCE);
    }

    #[test]
    fn test_syllable_distance_identical() {
        let a: Vec<String> = vec!["K", "AE1", "T"].iter().map(|s| s.to_string()).collect();
        assert_eq!(syllable_distance(&a, &a), 0);
    }

    #[test]
    fn test_syllable_distance_different_length() {
        let a: Vec<String> = vec!["K", "AE1", "T"].iter().map(|s| s.to_string()).collect();
        let b: Vec<String> = vec!["K", "AE1"].iter().map(|s| s.to_string()).collect();
        assert!(syllable_distance(&a, &b) > 0); // missing phoneme penalty
    }

    #[test]
    fn test_normalize_phoneme_ipa_vowel() {
        assert_eq!(normalize_phoneme("æ"), "AE");
        assert_eq!(normalize_phoneme("ɪ"), "IH");
    }

    #[test]
    fn test_normalize_phoneme_ipa_consonant() {
        assert_eq!(normalize_phoneme("k"), "K");
        assert_eq!(normalize_phoneme("ʃ"), "SH");
    }

    #[test]
    fn test_normalize_phoneme_already_arpabet() {
        assert_eq!(normalize_phoneme("AE1"), "AE1");
        assert_eq!(normalize_phoneme("K"), "K");
    }

    #[test]
    fn test_normalize_phoneme_diphthong() {
        assert_eq!(normalize_phoneme("aɪ"), "AY");
        assert_eq!(normalize_phoneme("oʊ"), "OW");
    }

    #[test]
    fn test_strip_stress() {
        assert_eq!(strip_stress("AE1"), "AE");
        assert_eq!(strip_stress("K"), "K");
        assert_eq!(strip_stress("IY0"), "IY");
    }
}
